modules/localization/msf/common/util/frame_transform.cc

```diff
@@ -0,0 +1,507 @@
+  double ep2 = 0.0;
+  double t = 0.0;
+  double t2 = 0.0;
+  double l = 0.0;
+  double l3coef = 0.0;
+  double l4coef = 0.0;
+  double l5coef = 0.0;
+  double l6coef = 0.0;
+  double l7coef = 0.0;
+  double l8coef = 0.0;
+  double tmp = 0.0;
+
+  /* Precalculate ep2 */
+  ep2 = (pow(sm_a, 2.0) - pow(sm_b, 2.0)) / pow(sm_b, 2.0);
+
+  /* Precalculate nu2 */
+  nu2 = ep2 * pow(cos(phi), 2.0);
+
+  /* Precalculate nn */
+  nn = pow(sm_a, 2.0) / (sm_b * sqrt(1 + nu2));
+
+  /* Precalculate t */
+  t = tan(phi);
+  t2 = t * t;
+  tmp = (t2 * t2 * t2) - pow(t, 6.0);
+
+  /* Precalculate l */
+  l = lambda - lambda0;
+
+  /* Precalculate coefficients for l**nn in the equations below
+   so a normal human being can read the expressions for easting
+   and northing
+   -- l**1 and l**2 have coefficients of 1.0 */
+  l3coef = 1.0 - t2 + nu2;
+
+  l4coef = 5.0 - t2 + 9 * nu2 + 4.0 * (nu2 * nu2);
+
+  l5coef = 5.0 - 18.0 * t2 + (t2 * t2) + 14.0 * nu2 - 58.0 * t2 * nu2;
+
+  l6coef = 61.0 - 58.0 * t2 + (t2 * t2) + 270.0 * nu2 - 330.0 * t2 * nu2;
+
+  l7coef = 61.0 - 479.0 * t2 + 179.0 * (t2 * t2) - (t2 * t2 * t2);
+
+  l8coef = 1385.0 - 3111.0 * t2 + 543.0 * (t2 * t2) - (t2 * t2 * t2);
+
+  /* Calculate easting (x) */
+  xy->x = nn * cos(phi) * l
+          + (nn / 6.0 * pow(cos(phi), 3.0) * l3coef * pow(l, 3.0))
+          + (nn / 120.0 * pow(cos(phi), 5.0) * l5coef * pow(l, 5.0))
+          + (nn / 5040.0 * pow(cos(phi), 7.0) * l7coef * pow(l, 7.0));
+
+  /* Calculate northing (y) */
+  xy->y = arclength0f_meridian(phi)
+          + (t / 2.0 * nn * pow(cos(phi), 2.0) * pow(l, 2.0))
+          + (t / 24.0 * nn * pow(cos(phi), 4.0) * l4coef * pow(l, 4.0))
+          + (t / 720.0 * nn * pow(cos(phi), 6.0) * l6coef * pow(l, 6.0))
+          + (t / 40320.0 * nn * pow(cos(phi), 8.0) * l8coef * pow(l, 8.0));
+}
+
+/*
+ * mapxy_to_latlon
+ *
+ * Converts x and y coordinates in the Transverse Mercator projection to
+ * a latitude/longitude pair.  Note that Transverse Mercator is not
+ * the same as UTM; a scale factor is required to convert between them.
+ *
+ * Reference: Hoffmann-Wellenhof, B., Lichtenegger, H., and Collins, J.,
+ *   GPS: Theory and Practice, 3rd ed.  New York: Springer-Verlag Wien, 1994.
+ *
+ * Inputs:
+ *   x - The easting of the point, in meters.
+ *   y - The northing of the point, in meters.
+ *   lambda0 - Longitude of the central meridian to be used, in radians.
+ *
+ * Outputs:
+ *   philambda - A 2-element containing the latitude and longitude
+ *               in radians.
+ *
+ * Returns:
+ *   The function does not return a value.
+ *
+ * Remarks:
+ *   The local variables Nf, nuf2, tf, and tf2 serve the same purpose as
+ *   N, nu2, t, and t2 in MapLatLonToXY, but they are computed with respect
+ *   to the footpoint latitude phif.
+ *
+ *   x1frac, x2frac, x2poly, x3poly, etc. are to enhance readability and
+ *   to optimize computations.
+ *
+ */
+void mapxy_to_latlon(double x, double y, double lambda0, WGS84Corr *philambda) {
```

Comment:
use const for inputs, check philambda before dereferencing.

Reply:
changed